interruptible compare.c

gasche opened the pull request:
In #12039, @eutro proposed a fix for issue #3921 (which is a more pressing issue in a multi-domain world) as follows:
- regularly poll for GC interrupts inside the compare loop
- if an interrupt is present, service the GC request and continue

In #12039, what is meant by "continue" is actually "restart the comparison from the start", which makes sense because the intermediate data-structure used by the comparison loop, the "compare stack", has been invalidated by a GC -- it contain interior pointers.

In #12053 we discussed changing the structure of the compare stack to avoid interior pointers. @eutro further suggested in the discussion in #12039 to use only valid OCaml values in the compare stack, that is to use tagged integers for block offsets and sizes.

The present PR combines the following:
- it uses a representation of the compare stack with only valid OCaml values
- it contains @eutro's original logic for regularly polling in the comparison function
- when a GC needs to be serviced, the compare stack is declared as a block of local roots

Note that in #12039 the polling logic is now more sophisticated than it was in the first iteration of the PR. I have only included the simple logic from the first commit, not the more complex code. I think that we should refine that (there are issues with the simple logic, some which we probably want to fix, in particular the looping on cyclic blocks of size 1), but the idea is to get the simplest overall code to demonstrate the combined design.

(My plan if we gather positive feedback on this approach would be to let @eutro adopt this approach if interested, and close the present PR.)

cc people who participated to discussions @eutro @smuenzel @xavierleroy @gadmm

### Performance impact

I re-ran the microbenchmark of #12053 (but remember that @smuenzel demonstrated that the results can be fairly different depending on the CPU), comparing the following versions:

- `trunk`: the current trunk. The compare stack uses interior pointers and the number of arguments left.
- `compare-root-stack-val-val-long`: the code of #12053 rebased on top of the current trunk;
  The compare stack uses base pointers and an offset, compared against the value size as recomputed from the value header.
  Benchmarking by @smuenzel has shown that this version is 10% faster or 10% slower than trunk depending on the CPU/machine.
- `compare-root-stack-val-val-long-long`: the variant of #12053 with base pointers, an offset and a size.
  Uses more memory, but seems consistently faster than trunk on all machines (for this test).
- `compare-root-stack-val-val-val-val`: using OCaml tagged integers for the offset and size.
  You will see below that it has the exact same performance as the `val-val-long-long` version.
- `compare-interruptible`: adding the polling logic

Benchmark results:

```
Benchmark 1: ./ocamlrun.trunk test.byte --width 7 --height 7 --niter 10
  Time (mean ± σ):     454.1 ms ±   9.6 ms    [User: 451.0 ms, System: 1.4 ms]
  Range (min … max):   447.7 ms … 508.9 ms    50 runs
 
Benchmark 2: ./ocamlrun.compare-root-stack-val-val-long test.byte --width 7 --height 7 --niter 10
  Time (mean ± σ):     406.2 ms ±   2.9 ms    [User: 403.1 ms, System: 1.7 ms]
  Range (min … max):   404.0 ms … 417.0 ms    50 runs
 
Benchmark 3: ./ocamlrun.compare-root-stack-val-val-long-long test.byte --width 7 --height 7 --niter 10
  Time (mean ± σ):     390.3 ms ±   5.8 ms    [User: 387.1 ms, System: 1.7 ms]
  Range (min … max):   381.1 ms … 406.1 ms    50 runs
 
Benchmark 4: ./ocamlrun.compare-root-stack-val-val-val-val test.byte --width 7 --height 7 --niter 10
  Time (mean ± σ):     389.9 ms ±   4.6 ms    [User: 386.9 ms, System: 1.6 ms]
  Range (min … max):   386.2 ms … 411.1 ms    50 runs
 
Benchmark 5: ./ocamlrun.compare-interruptible test.byte --width 7 --height 7 --niter 10
  Time (mean ± σ):     411.5 ms ±   4.1 ms    [User: 408.4 ms, System: 1.6 ms]
  Range (min … max):   405.2 ms … 427.6 ms    50 runs
 
Summary
  './ocamlrun.compare-root-stack-val-val-val-val test.byte --width 7 --height 7 --niter 10' ran
    1.00 ± 0.02 times faster than './ocamlrun.compare-root-stack-val-val-long-long test.byte --width 7 --height 7 --niter 10'
    1.04 ± 0.01 times faster than './ocamlrun.compare-root-stack-val-val-long test.byte --width 7 --height 7 --niter 10'
    1.06 ± 0.02 times faster than './ocamlrun.compare-interruptible test.byte --width 7 --height 7 --niter 10'
    1.16 ± 0.03 times faster than './ocamlrun.trunk test.byte --width 7 --height 7 --niter 10'
```

On my machine we see that the polling logic introduces a non-neglectible cost, but that the PR remains faster than trunk. (I believe that this comes from the fact that we only increment one offset instead of two interior pointers.)

gasche committed:
compare.c: avoid inner pointers on the compare stack

To compare same-size blocks containing values, do_compare_val recursively
compares the first argument of each block, and pushes "the rest of the
arguments' on both sides to the compare stack. The compare stack thus
contains pairs of sequences of arguments to be compared.

Before this PR, pairs of sequences of arguments are represented by:

- two inner pointers (into the parent blocks) to the first arguments
  to be compared on each side
- a "count" field remembering the number of arguments still to compare

The use of inner pointers that are not valid OCaml values makes some
things delicate, for example we could not ask the GC to scan the
compare stack (as is being considered in the context of # 12039).

After this PR, pairs of sequences of arguments are represented by:

- pointers to the parent containing blocks on each side
  (that is, valid OCaml values)
- an "offset" field remembering the offset of the next argument to
  compare, stored as a tagged OCaml integer
- a "size" field remembering the size of the two blocks,
  store as a tagged OCaml integer

At this point, the compare stack contains only valid OCaml values and
could be registered as a single block of local GC roots.

Note: storing the size is superfluous as the size can also be computed
from the blocks themselves (bit-shifting on the header value). In our
tests, storing the size and comparing directly against it noticeably
improves performance on some architectures (a 10-20% different on
compare microbenchmarks). The cost is to use 4 words per element
instead of 3.

gasche committed:
Poll for interrupts in structural equality loop

Co-authored-by: eutro <git@eutro.dev>
Co-authored-by: Xavier Leroy <xavier.leroy@college-de-france.fr>

gasche committed:
compare.c: minor fix in stack-handling code

The previous stack-push code would increment the stack pointer before
writing the new element, so the very first slot was never used.

This is slightly inefficient and makes registering the compare stack
as GC roots weird (we would have to start at 1).

gadmm commented on runtime/compare.c:
```diff
@@ -293,16 +297,33 @@ static intnat do_compare_val(struct compare_stack* stk,
+        exn = caml_do_pending_actions_exn();
+        End_roots();
+        if (Is_exception_result(exn)) {
+          caml_raise(exn);
```
There is a missing `Extract_exception`; or simply call `caml_raise_if_exception`.

gadmm replied:
Or simpler for what you wrote: call `caml_process_pending_actions`. But I suspect you also want to clear the stack before raising.

gasche replied:
Good catch (we missed this issue when reading the code of #12039), I added a `compare_free_stack` call before the `caml_raise`.

gadmm replied:
Where's my `Extract_exception` ? :)

gasche replied:
Oops, re-done.

gadmm commented:
Have you tried to investigate where the overhead of the polling logic comes from? Does the value of `COMPARE_SIGNAL_POLL_PERIOD` have an impact? If the micro-benchmark is really that sensitive to incrementing one pointer instead of two, it might also be the case that the effect of code layout is important as well. I'd be wary about drawing conclusions such as "polling costs 5%".

gasche commented:
> Does the value of COMPARE_SIGNAL_POLL_PERIOD have an impact?

No, I tried making it 32 times higher in the hope of reducing overhead, and it was in the noise.

eutro commented on runtime/compare.c:
```diff
@@ -280,9 +283,10 @@ static intnat do_compare_val(struct compare_stack* stk,
+        sp->offset = Val_long(1);
+        sp->size = Val_long(sz1);
       }
       /* Continue comparison with first field */
```
Recall that the `continue;` after this needs to instead be a jump to the poll block to ensure that the comparison in e.g.
```ocaml
type nat = S of nat | Z;;
let rec inf = S inf;;
inf = inf;;
```
polls.

eutro replied:
It looks like you've moved it anyway.

gasche replied:
I moved the poll check to the entry of the `while` loop as an easy way to fix this issue. I haven't tried (yet) to integrate your code to run regularly inside bounded loops, because I find it a bit too complex to my taste.

gasche commented:
>  it might also be the case that the effect of code layout is important as well

Yep, so in fact I added a `CAMLunlikely` hint on the poll check and the overhead of polling is gone entirely, it now has the same performance than the other versions such as val-val-val-val.

~~(I moved the poll check to the entry of the loop to ensure that it does run regularly.)~~

eutro commented on runtime/compare.c:
```diff
@@ -118,9 +136,16 @@ static intnat do_compare_val(struct compare_stack* stk,
+       unbounded time. */
+    if (CAMLunlikely(++signal_poll_timer >= COMPARE_SIGNAL_POLL_PERIOD)) {
+      signal_poll_timer = 0;
+      poll_pending_actions(stk);
```
It seems like `v1` and `v2` are kept live across this without being registered as roots.

gasche replied:
Right... I moved the poll logic to take place just before the `next_item` label, at this point `v1, v2` are not kept live as they will be reloaded from the compare stack.

gasche replied:
(It would also be possible to keep the poll test at the top and root v1, v2 at this point, but doing so requires to copy them into an array to not pessimize the code, and even with that th result is noticeably slower on my machine.)

gadmm replied:
Another option would be to pass the addresses of v1 and v2 and treat them as arrays of size one for `Begin_roots_block`.

gasche replied:
Any solution that involves registering v1, v2 as roots or taking their address is substantially slower on my machine; I suspect that the compiler was able to put them in registers before.

gadmm replied:
Right, the compiler cannot optimize the spilling/reloading as initially thought. However, by manually copying/restoring v1 and v2 within the inner block, and rooting the copies as suggested, one gets a performance within the irrelevant code layout effects. Strangely, copying into an array instead of two variables and rooting the array as you did is not equivalent, it comes with a huge cost.

That being said, I got accustomed to the more complicated version, so feel free to choose what works best for you.

gadmm commented:
> Yep, so in fact I added a `CAMLunlikely` hint on the poll check and the overhead of polling is gone entirely, it now has the same performance than the other versions such as val-val-val-val.

I suspect that there is a chance factor in your new result. On my setup I had to move the poll to a separate function and give it the attribute noinline to observe a gain in performance (you did the former but not the latter). Even then, it only removed half of the overhead.

I gained a further 1-2% (unrelated to polling) by optimizing the code layout of the rest of the function, it would have been for another PR but I am not sure it is worth the added complexity.

gadmm commented:
Still investigating this overhead, I have tried to reproduce your new result on my machine and I found that it now goes _slower_ than it used to, around 9-10% compared to not polling (on mine). Then I made the following change and found that I am back to 1-2%:
```diff
diff --git a/runtime/compare.c b/runtime/compare.c
index c3f04849fc..cc0d7239d4 100644
--- a/runtime/compare.c
+++ b/runtime/compare.c
@@ -32,7 +32,7 @@ struct compare_item { value v1, v2, offset, size; };
 #define COMPARE_STACK_MAX_SIZE (1024*1024)
 
 #define COMPARE_SIGNAL_POLL_PERIOD 1024
-#define COMPARE_MUST_POLL(timer) (CAMLunlikely(timer >= COMPARE_SIGNAL_POLL_PERIOD))
+#define COMPARE_MUST_POLL(timer) (CAMLunlikely(timer % COMPARE_SIGNAL_POLL_PERIOD == 0))
 
 struct compare_stack {
   struct compare_item init_stack[COMPARE_STACK_INIT_SIZE];
@@ -101,6 +101,7 @@ static intnat compare_val(value v1, value v2, int total)
   return res;
 }
 
+__attribute__((noinline))
 static void poll_pending_actions(struct compare_stack* stk)
 {
   if (caml_check_pending_actions()) {
@@ -136,7 +137,7 @@ static intnat do_compare_val(struct compare_stack* stk,
 {
   struct compare_item * sp;
   tag_t t1, t2;
-  int signal_poll_timer = 0;
+  unsigned int signal_poll_timer = 0;
 
   sp = stk->stack;
   while (1) {
@@ -327,7 +328,6 @@ static intnat do_compare_val(struct compare_stack* stk,
       /* At this point v1, v2 need not be rooted, they are about to be
          reloaded from the compare stack below. */
       poll_pending_actions(stk);
-      signal_poll_timer = 0;
     }
 
   next_item:
```
Using `test` is better than using `cmp` but the change to use modulo should not have such a large impact, this is probably again weird code layout effects.

gasche commented:
So the polling versions:
- my original placement of the poll was wrong for two reasons, a "known" reason that it would not poll on size-1 blocks and an unknown-to-me issue about the fact that it was at a place where v1, v2 should be kept live; that original wrong version was somewhat slower as-is and exactly-as-fast (as my val-val-val-val version without polls) with the CAMLunlikely hint and the code moved to a separate function
- I did not find a way to keep v1, v2 live at that point without hurting performance too much, so I moved to a different strategy where we always put all arguments on the stack when we must poll, and can thus poll at a place where v1, v2 are not live anymore; this version is indeed somewhat slower than my val-val-val-val version on my machine, but still faster (on my machine) than the code in trunk

The bad news is that the testsuite appears to fail (on my new old friend intext_par), so there is something wrong with the PR. Finding and fixing the bug is higher-priority for me than any micro-optimization discussion.

In the medium term (assuming the bug is fixed), I think that a reasonable target for this PR would be to be "not measurably slower than trunk" on a robust set of test machine (I hope @smuenzel will be willing to help with that). Once we reach this I would vote to declare victory and look at other things than micro-optimizing the polling formulation.

(One idea I had in mind -- for *another* PR -- was to introduce a prefetching queue in this algorithm. I wonder if this can be done easily (I don't see why it would be harder than in the GC itself, but maybe I am missing something), and if it works the benefits may trump the weird code-placement effects.(

gadmm commented:
Yes, one should not care about these layout effects. If one cares about performance, it is better to first address memory effects.

smuenzel commented:
> In the medium term (assuming the bug is fixed), I think that a reasonable target for this PR would be to be "not measurably slower than trunk" on a robust set of test machine (I hope @smuenzel will be willing to help with that). Once we reach this I would vote to declare victory and look at other things than micro-optimizing the polling formulation.

No problem, I will get to this soon

smuenzel commented on runtime/compare.c:
```diff
@@ -118,9 +136,11 @@ static intnat do_compare_val(struct compare_stack* stk,
 {
   struct compare_item * sp;
   tag_t t1, t2;
+  int signal_poll_timer = 0;
```
counting down from `COMPARE_SIGNAL_POLL_PERIOD` is also an option. Then all the comparisons will be against 0. (this has no measurable impact, FWIW)

gasche replied:
On my machine counting down to 0 has a measurable impact, and the code is equally clear, so I went for it. (At the time of writing, the previous head with polling was 10% slower than the version without any polling logic, and the downto-0 version is only 5% slower.)

smuenzel commented on runtime/compare.c:
```diff
@@ -99,6 +101,22 @@ static intnat compare_val(value v1, value v2, int total)
+  if (caml_check_pending_actions()) {
+    value exn;
+    Begin_roots_block((value*)(stk->stack),
+                      (stk->limit - stk->stack) * sizeof(struct compare_stack) / sizeof(value));
```
Are all these values valid ocaml values? Isn't it just from `stk->stack` to `sp`?

smuenzel replied:
Regarding your bug: even after fixing the roots block size, I still see some `0` roots in the roots block on the failing tests.

gasche replied:
Actually, I wonder if there may be an existing bug/glitch/weirdness in the indexing -- or maybe this is just the lack of sleep on my side hallucinating bugs, apologies. We would naively expect one of these invariants to hold, but which one?

1. the slots from `stk->stack` to `sp` *excluded* contain valid stack slots
2. the slots from `stk->stack` to `sp` *included* contain valid stack slots

My reading of the code is that we expect `sp` to point to the "current" stack slot, and we increment `sp` before filling a new slot, so "included" interpretation (2), but in fact we initialize `sp` to `stk->stack` without pushing an element at the beginning, so "excluded" interpretation (1) at initialization.

So my understanding is that in fact, `stk->stack` is allocated memory that plays the role of a sentinel value, and is never actually filled with any valid slack slot. We only ever dereference `sp` after incrementing it (so we are working on the non-zero slot) or after checking the sp value against the sentinel value, so this is correct, but we are missing one slot of our 8 initial slots, so this may limit performance slightly by causing reallocation earlier than necessary.

I need to fix the root range to stop at `sp` (thanks for this remark @smuenzel that set me on the right track), but for now I also need to have it actually start at `sp->stack + 1`.

If my analysis above is correct, I would consider removing the sentinel slot to get a simpler structure and obtain an extra slot.

gasche replied:
(see other threads)

gasche replied:
Note: I went back to read the 4.07 version of compare.c which uses the same logic for `sp` with a surprise sentinel value; this is not a recent change related to Multicore (in 5.x or buildup work).

smuenzel commented on runtime/compare.c:
```diff
@@ -99,6 +101,22 @@ static intnat compare_val(value v1, value v2, int total)
+  if (caml_check_pending_actions()) {
+    value exn;
+    Begin_roots_block((value*)(stk->stack),
+                      (sp - stk->stack) * sizeof(struct compare_stack) / sizeof(value));
```
should this be `sizeof(struct compare_item)` rather than `sizeof(struct compare_stack)`?

smuenzel replied:
After that "Test 425" in `intext_par` still fails, not really sure why yet.

gasche replied:
Great, I tried to denote a portion of linear memory and all three of the start index, end index and element size were wrong. Thanks!

gasche commented:
Thanks to @smuenzel's review and some extra head scratching (one line had three independent bugs, one of them subtle), intext_par.ml now passes on my machine.

smuenzel commented:
I have a slightly different proposal (that also passes the test). Here, `sp` is renamed `next_stack_slot`:
https://github.com/gasche/ocaml/compare/compare-interruptible...smuenzel:compare-interruptible

smuenzel commented:
Even if that's not completely correct, I would vote for more descriptive variable names in general, the intended usage of `sp` was definitely a little hard to understand.

gasche commented:
@smuenzel except for your debatable taste in writing `struct compare_item*sp` (is this out of annoyance with the debate on where to place the space?), I like your commit and would be willing to include it in the PR. But could you also change the argument name in the polling function for consistency?

smuenzel commented:
Updated, and I made some of the long lines shorter.

xavierleroy commented:
>  (one line had three independent bugs, one of them subtle)

The C language is a harsh mistress, as the saying goes...

I haven't reviewed yet, and it may take a while before I can.  Still, I agree that `v1` and `v2` should not be registered as roots for best performance.  The intent of these variables is to cache the top of the compare stack in registers, and GC roots cannot reside in registers.

smuenzel commented:
Some benchmarks (using gcc 12.2.0, native ocaml with flambda)

<table>
<tr>
	<td> Machine
	<td> Microarchitecture
	<td> trunk (2231b5e83c)
	<td> pr12128 (3f363a4297)
<tr>
	<td> Intel i7-11800H
	<td> Willow Cove (2020)
	<td> 137.4 ms ±   3.6 ms
	<td> 117.9 ms ±   3.5 ms
<tr>
	<td> AMD Ryzen 7 5800X
	<td> Zen 3 (2020)
	<td> 117.9 ms ±   1.5 ms
	<td> 109.8 ms ±   1.6 ms
<tr>
	<td> AMD Ryzen 7 3700X
	<td> Zen 2 (2019)
	<td> 142.1 ms ±   1.9 ms
	<td> 136.8 ms ±   2.5 ms
</table>

In each case, this PR is faster than trunk.

gasche commented:
My current thinking on this PR:

- We have fixed the obvious bugs, and reached the performance target "not slower than trunk" *on my microbenchmark*.
- Regarding correctness, I think that this is ready for a review, so I will turn it from Draft to review-ready.
- Regarding performance, I wonder if we could get Sandmark benchmarks to confirm the not-slower performance target on a wider variety of benchmarks (in particular involving small/short comparisons instead of only testing long/slow comparisons)
- I would also like to cleanup the commit history to merge some of the incremental changes done during yesterday's collaborative effort.

gasche commented:
Ah, also: Initially I proposed to integrate this code back in #12039. I am still open to the idea if @eutro is willing, but my impression is that these changes are tricky and that making them larger / more complex is not necessarily a good idea for reviewability. So maybe it makes sense to try to get the current merge with the current feature scope (that is: polling purely on the block count, without any specific logic for polling inside large blocks), and have follow-up PRs to refine polling as done in #12039.

eutro commented:
I'd agree to merging this PR with its current feature set.

I'd even be slightly opposed to actually performing polling during the string/float array comparison loops due to the added complexity: as @xavierleroy said, "string comparison will terminate eventually, and rather quickly even for very big strings" (and the same holds for float arrays). I think the most we should do is decrement the timer by more than just one on large strings/float arrays, to ensure good enough latency. However, that discussion should still be part of another PR.

xavierleroy commented on runtime/compare.c:
```diff
@@ -99,6 +100,31 @@ static intnat compare_val(value v1, value v2, int total)
+{
+  if (caml_check_pending_actions()) {
+    value exn;
+    if (next_stack_slot == stk->stack) {
+      exn = caml_do_pending_actions_exn();
+    } else {
```
Zero-sized blocks of local roots work fine as far as I can see.  (Function `caml_do_local_roots` in runtime/roots.c)  So you don't need this smelly special case, the general case below should handle it just fine.

gasche replied:
Thanks, fixed.

xavierleroy commented on runtime/compare.c:
```diff
@@ -99,6 +100,31 @@ static intnat compare_val(value v1, value v2, int total)
+      size_t roots_length =
+        (next_stack_slot - stk->stack)
+        * sizeof(struct compare_item) / sizeof(value);
+      Begin_roots_block(roots_start, roots_length);
```
+1 for using the old roots API, it works just right here.  (And elsewhere.)  @damiendoligez is going to grumble about it, though.

xavierleroy commented on runtime/compare.c:
```diff
@@ -116,11 +142,13 @@ static intnat compare_val(value v1, value v2, int total)
                              value v1, value v2, int total)
 {
-  struct compare_item * sp;
+  struct compare_item * next_stack_slot;
```
Renaming `sp` into `next_stack_slot` doesn't help.  A stack pointer is ALWAYS named `sp`, and if the stack grows upward, which is the case here, it ALWAYS points just above the latest stack entry, so that pushing `v` is `*sp++ = v` and popping `v` is `v = *--sp`.  These conventions are ingrained in my mind and in that of many systems programmers, so please follow them instead of inventing your own names.

gasche replied:
I'm happy to revert this part of the change. Can you confirm that there was indeed a slight issue with the original code, in the sense that the first stack slot was never written to? The fix is to ensure that we write the value on push before incrementing the stack pointer, when previously it would increment first.

xavierleroy replied:
This is embarrassing, even though I did not write the old code: yes, the old code does not follow the `sp` convention I have in mind...  I agree with your fix, but would prefer `next_stack_slot` to be named `sp` and `sp` (in the line `struct compare_item* sp = next_stack_slot-1;`) to be named `item` or something like that.  Thanks for your patience.

smuenzel replied:
I'm not familiar with the convention regarding stacks and the `sp` variable, do we have that documented somewhere?
Having the code be (semi-)self-documenting is very helpful for those joining the codebase later, and helps with conveying the intentions in review. Naming the variable according to its intended usage (`next_stack_slot`) is a very easy way to accomplish that.

So maybe we can keep `next_stack_slot`, and call the item `item`? I agree that `item` is a much better name than `sp` in the context of the `compare_item` line

gasche replied:
I am planning to rename `next_stack_slot` back to `sp` to follow the implicit compiler-backend convention, and name the item `item`. Let's not waste our time discussing the details here when a code owner expresses a firm opinion.

gasche replied:
I renamed `next_stack_slot` back to `sp`. I use `next` as the item name when we are pushing a new item, and `last` as the item name when we are accessing the top of the stack.

xavierleroy commented on runtime/compare.c:
```diff
@@ -276,33 +304,57 @@ static intnat do_compare_val(struct compare_stack* stk,
-        sp->v2 = &Field(v2, 1);
-        sp->count = sz1 - 1;
+      /* Remember that we still have to compare fields 1 ... sz - 1. */
+      if (sz1 > 1 || CAMLunlikely(signal_poll_timer == 0)) {
```
These `signal_poll_timer == 0` comparisons are fragile.  If `signal_poll_timer` can be decremented several times before testing, you can miss the `0` value and get a negative `signal_poll_timer`, which will test negative for 2^31 more iterations or so.  This can happen in the `Forward_tag` case, which `continue`s straight to the beginning of the `while (1)` loop, decrementing `signal_poll_timer` a second time.  Please write robust tests `signal_poll_timer <= 0`.

gasche replied:
Done, thanks.

(I thought about changing those `continue` to `goto continue_after_tick` to avoid this, because using `== 0` is neater, but I went for the more robust test.)

xavierleroy replied:
If you want a `signal_poll_timer == 0` test,  decrement `signal_poll_timer` just before the test, not two pages above, so that it is obvious that the timer cannot skip value 0.

gasche replied:
Yes, something like `if (--poll_timer == 0) must_poll = 1;` as the tick instruction, and then checking the `must_poll` variable as the poll test. But I am content with `<= 0`.

xavierleroy commented on runtime/compare.c:
```diff
@@ -276,33 +304,57 @@ static intnat do_compare_val(struct compare_stack* stk,
+           whole blocks to the compare stack, and break (the switch)
+           to the poll point below instead of continuing at the
+           top. */
+        if (next_stack_slot >= stk->limit)
+          next_stack_slot = compare_resize_stack(stk, next_stack_slot);
+        struct compare_item* sp = next_stack_slot;
+        next_stack_slot ++;
+        sp->v1 = v1;
+        sp->v2 = v2;
+        sp->size = Val_long(sz1);
+        if (CAMLunlikely(signal_poll_timer == 0)) {
+          sp->offset = Val_long(0);
+          break;
+        } else {
+          sp->offset = Val_long(1);
+        }
```
It seems you worked hard to combine two cases of pushing to the stack in one piece of code.  One case is when we need to descend into a pair of blocks, and the other is when we need to poll.  My instinct is to keep both cases distinct.  Before I suggest another code layout, can you tell me why you chose to combine these two cases of pushing?

gasche replied:
I assumed that code size could be performance-sensitive for this function, and that sharing most of the code of the two cases was better. I am happy to separate the two cases as you suggest, but the obvious way I can think of would duplicate the `push` logic.

xavierleroy replied:
OK, I'll keep that in mind and think about it.

xavierleroy reviewed (commented):
This is making good progress, thanks.  See below for some suggestions and a question concerning the general organization of the code.

Tip for reviewers: the changes are difficult to understand with the Github GUI, I found it easier to fetch this branch and compare using Emacs ediff.

Tip for people who run benchmarks: I like to measure how many instructions are executed (under Linux, use `perf stat -e instructions command arguments`).  This number does not predict execution times exactly (but the correlation is good), however it is highly reproducible.

gasche commented:
> Tip for people who run benchmarks: I like to measure how many instructions are executed (under Linux, use `perf stat -e instructions command arguments`). This number does not predict execution times exactly (but the correlation is good), however it is highly reproducible.

It looks like we are in a situation where the instruction count and the compute times do not correlate. Comparing `trunk` with `CUR` (the current state of the branch), on my machine CUR is faster with more instructions, but less cycles. My understanding is that that it comes from more well-predicted checks (and decrements), and one less memory write in the hot path.

```
$ perf stat -e instructions,cycles ./ocamlrun.trunk ./test.byte --width 8 --height 8 --niter 10
    40,463,289,528      instructions:u                   #    2.54  insn per cycle         
    15,916,988,671      cycles:u                                                           
       8.900625915 seconds time elapsed

$ perf stat -e instructions,cycles ./ocamlrun.CUR ./test.byte --width 8 --height 8 --niter 10
    46,065,619,535      instructions:u                   #    3.19  insn per cycle         
    14,446,607,666      cycles:u                                                           
       8.081545183 seconds time elapsed
```

smuenzel commented:
Try also sampling `cache-misses`, `branches` and `branch-misses`, it might give more insight.

>  one less memory write in the hot path.

To first approximation, memory writes should be free, so I wouldn't expect a small change in memory writes to impact performance unless something else is also triggered by it.

gasche commented:
I believe that I have addressed all review comments so far.

I would rather not spend too much time analyzing `perf stat -d -d -d` output (I did it a bit already, nothing conclusive) unless there is a clear need. My goal is to enable a new feature (polling inside the comparison code) without a performance regression, and I think we have enough experimental evidence to conclude that there is no performance regression on this benchmark.

My understanding of potential remaining work items:

- the PR needs an approval, which presumably would come from someone (besides me) being content with the PR code
- I mentioned running a round of Sandmark benchmarks. I was too lazy to follow the process to do this so far. I think of it as a nice-to-have, maybe not necessary.
- @xavierleroy mentioned thinking about simplifying the push-before-poll logic in the code (there are now two logical paths where we push to the compare stack, one for >1-arguments block and one in must-poll conditinos). I think of it as a nice-to-have, maybe not necessary. (Xavier, if you think this is necessary, let me know and I will do the obvious thing of duplicating the code and measuring that there is no clear performance impact.)
- I need to rebase the PR to cleanup the history. This can be done at any time.

smuenzel commented on runtime/compare.c:
```diff
@@ -110,25 +110,20 @@ static intnat compare_val(value v1, value v2, int total)
-    }
+  if (!caml_check_pending_actions()) return;
+  value exn;
+  CAMLassert(sp > stk->stack);
```
You removed the special case `sp == stk->stack`, but it can still occur, it's just fine when it does, right? So I guess we need to change the assert.

gasche replied:
Oops! Indeed, the assert should be removed. I will fixup the relevant commit. Thanks.

xavierleroy replied:
The assert is still there?

smuenzel reviewed (changes requested):
Other than the small issue with the assert, and a changelog entry, I believe this pr is good to go.

gasche commented on runtime/compare.c:
```diff
@@ -99,6 +107,26 @@ static intnat compare_val(value v1, value v2, int total)
+  value* roots_start = (value*)(stk->stack);
+  size_t roots_length =
+    (sp - stk->stack)
+    * sizeof(struct compare_item) / sizeof(value);
```
I have a question for people who know C: could the line above

```c
  size_t roots_length =
    (sp - stk->stack)
    * sizeof(struct compare_item) / sizeof(value);
```

someone be obtained without explicit size calculation, by casting the pointers at a different type instead? Would something like the following work?

```c
  size_t roots_length = (value*)sp - (value*)(stk->stack);
```

(Reminder: `sp` and `stk->stack` are pointers into an array of structures containing four values.)

xavierleroy replied:
The other formula should work provided there is no padding in the `struct compare_item`, and if there is any the GC will probably crash...  For additional clarity you could write
```
value* roots_start = (value*)(stk->stack);
value* roots_end = (value *)sp;
size_t roots_length = roots_end - roots_start;
 ```

gadmm replied:
@gasche You can add an assertion if you are nervous about this layout hypothesis. (I looked for a way to make it a static assertion but I did not find a portable way to do so; offsetof does not work here, and remember that some platforms complain about the null pointer offset trick.)

smuenzel commented:
Your latest question about the c standard got me to think a bit about how we're accessing these structs (yes, it's undefined behaviour, but I think that's actually fine, since most c compilers are somewhat sane). What's not portable, however, is the amount of padding that the struct has -- the compiler can decide to add padding between (or after) the members, which would break our root block.

We should think about adding ``__attribute__((__packed__))`` to the struct, although I'm not sure how to do that in a portable way.

gasche committed:
Changes

xavierleroy commented:
Here is my proposal for a slightly different control flow: https://github.com/xavierleroy/ocaml/commit/0677dd6db0254e023fa403dae3080590b40c40fe (the branch is at https://github.com/xavierleroy/ocaml/tree/compare-interruptible ).  The overall structure is a `while (--poll_timer > 0) { ... }` followed by the polling logic.  Also, during polling, I don't push `v1` and `v2` on the compare stack, just save them in local variables registered as GC roots.

Some performance figures (on an aging Skylake Xeon):
```
Benchmark 1: ./eqbench.trunk --width 7 --height 7 --niter 10
  Time (mean ± σ):     186.1 ms ±   9.1 ms    [User: 185.0 ms, System: 1.1 ms]
  Range (min … max):   171.3 ms … 195.0 ms    50 runs
 
Benchmark 2: ./eqbench.gasche --width 7 --height 7 --niter 10
  Time (mean ± σ):     185.3 ms ±   9.8 ms    [User: 184.2 ms, System: 1.1 ms]
  Range (min … max):   171.6 ms … 195.1 ms    50 runs
 
Benchmark 3: ./eqbench.xl --width 7 --height 7 --niter 10
  Time (mean ± σ):     175.9 ms ±   8.5 ms    [User: 174.3 ms, System: 1.6 ms]
  Range (min … max):   160.2 ms … 186.7 ms    50 runs

Performance counter stats for './eqbench.trunk --width 7 --height 7 --niter 10':
     2 020 176 637      instructions                                                     1 434 392      branch-misses                                           
Performance counter stats for './eqbench.gasche --width 7 --height 7 --niter 10':
     2 214 834 064      instructions                                                     1 728 939      branch-misses                                           
Performance counter stats for './eqbench.xl --width 7 --height 7 --niter 10':
     2 194 591 570      instructions                                                     1 471 298      branch-misses                                           
```
No dramatic differences.

If you like measurements, you should try a  left-leaning structure such as
```
type 'a leftlist = Nil | Snoc of 'a leftlist * 'a
```
Comparison requires a deep stack, so it should exhibit memory effects coming from the fact that stack entries are now 4 words instead of 3.

xavierleroy commented:
> We should think about adding __attribute__((__packed__)) to the struct, although I'm not sure how to do that in a portable way.

If you're really worried about padding, just use `value [4]` as the type of stack items.  But according to the ELF ABI padding rules, there should be none in `struct compare_item`.  I'd rather not use the `packed` attribute.

damiendoligez reviewed (approved):
Green-checking on behalf of @smuenzel.

I had a look at the code and I grumbled in private about `Begin_roots`/`End_roots`. I'm happy with both things.

gasche commented:
TODO: I still need to think about Xavier's proposal for the rooting of v1, v2, and rebase the PR.

gasche commented:
In the end I decided to keep Xavier's code as-is, and I rebased the PR to squash the intermediate commits together. (There are now three independent commits, the one changing the compare stack representation, the minor fix to the stack-push code, and the polling logic.) The code is unchanged from the last review round, so this should be good to merge if the CI agrees.

xavierleroy reviewed (approved):
Note for reviewers: this is much  easier to review if you tick the "hide whitespace" box in the gear menu near "file filter" and "conversations".

Looks good to me, except one assertion that should probably be removed based on earlier discussions.

xavierleroy commented:
Also, check-typo is unhappy.

gasche commented:
I tried a benchmark with a left-leaning data structure. In fact I simply used my current benchmark of n-ary trees, tweaked to store the node value after the children.

```ocaml
type 'a tree = Node of 'a tree list * 'a

let rec make_perfect v ~width ~height =
  let children =
    if height = 0 then [] else
    let child = make_perfect v ~width ~height:(height - 1) in
    List.init width (fun _ -> child)
  in
  Node (children, v)

let width = ref 0
let height = ref 0
let niter = ref 0

let usage_msg = "Compares two perfect trees."
let spec = Arg.[
  "--width", Set_int width, "<int> the width of the generated trees";
  "--height", Set_int height, "<int> the height of the generated trees";
  "--niter", Set_int niter, "<int> number of comparison iterations";
]

let () = Arg.parse spec failwith usage_msg

let () =
  if !width = 0 || !height = 0 || !niter = 0 then begin
    Arg.usage spec usage_msg;
    exit 2
  end

let () =
  let t1 = make_perfect (42, 43, 44) ~width:!width ~height:!height in
  let t2 = make_perfect (42, 43, 44) ~width:!width ~height:!height in
  for _ = 1 to !niter do
    assert (t1 = t2)
  done
```

Running this program with parameters `--height 240_000 --width 1 --niter 10` suggests that the current version is 5% faster than trunk on my machine: we haven't been missing a noticeable slowdown due to the larger memory footprint of the 4-word stack elements. (Higher `height` parameter fail with an exception as they overflow the hard limit on the compare stack, unchanged from trunk.)

eutro commented:
Sorry, could my full first name not be included in the changelog please? Preferably not in the commit history either.
